src/main/java/com/mixpanel/android/viewcrawler/ViewVisitor.java

```diff
@@ -152,19 +190,120 @@ public void accumulate(View found) {
             }
-            setLayout(found, newVerb, newAnchorId);
+
+            try {
+                setLayout(found, newVerb, newAnchorId);
+            } catch (CantVisitException e) {
+                cleanup();
+                mOnEditErrorListener.sendErrorMessage(e);
+            }
         }
 
-        private void setLayout(View target, int verb, int anchorId) {
+        private void setLayout(View target, int verb, int anchorId) throws CantVisitException {
             RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)target.getLayoutParams();
             params.addRule(verb, anchorId);
+
+            final Set<Integer> rules;
+            if (mHorizontalRules.contains(verb)) {
+                rules = mHorizontalRules;
+            } else if (mVerticalRules.contains(verb)) {
+                rules = mVerticalRules;
+            } else {
+                rules = null;
+            }
+
+            if (rules != null && !verifyLayout(target, rules)) {
+                throw new CantVisitException("Circular dependency detected!", "circular_dependency", mName);
+            }
+
             target.setLayoutParams(params);
         }
 
+        private boolean verifyLayout(View target, Set<Integer> rules) {
+            ViewGroup parent = (ViewGroup) target.getParent();
+            SparseArray<View> idToChild = new SparseArray<View>();
+
+            int count = parent.getChildCount();
+            for (int i = 0; i < count; i++) {
+                View child = parent.getChildAt(i);
+                int childId = child.getId();
+                if (childId > 0) {
+                    idToChild.put(childId, child);
+                }
+            }
+
+            ArrayMap<View, ArrayList<View>> dependencyGraph = new ArrayMap<View, ArrayList<View>>();
+            int size = idToChild.size();
+            for (int i = 0; i < size; i++) {
+                final View child = idToChild.valueAt(i);
+                final RelativeLayout.LayoutParams childLayoutParams = (RelativeLayout.LayoutParams) child.getLayoutParams();
+                int[] layoutRules = childLayoutParams.getRules();
+
+                ArrayList<View> dependencies = new ArrayList<View>();
+                for (int rule : rules) {
+                    int dependencyId = layoutRules[rule];
+                    if (dependencyId > 0 && dependencyId != child.getId()) {
+                        dependencies.add(idToChild.get(dependencyId));
+                    }
+                }
+
+                dependencyGraph.put(child, dependencies);
+            }
+
+            return hasCycle(dependencyGraph);
+        }
+
+        private boolean hasCycle(ArrayMap<View, ArrayList<View>> dependencyGraph) {
+            ArrayList<View> dfsStack = new ArrayList<View>();
+            while (!dependencyGraph.isEmpty()) {
+                View currentNode = dependencyGraph.keyAt(0);
+                if (!subGraphHasCycle(dependencyGraph, currentNode, dfsStack)) {
```

Comment:
We may want to rename subgraphHasCycle, since in addition to returning a value it has a side effect. Perhaps `detectSubgraphCycle`? Something with a verb in it?